reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -19,610 +19,1029 @@
+			}
+			if (key == Attr.RUN_STYLE) {
+				return Attr.RunStyle.SYNC;
+			}
+
+			return InnerOperator.super.scanUnsafe(key);
+		}
+
+		@Override
+		public CoreSubscriber<? super R> actual() {
+			return this.outer;
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (Operators.validate(this.s, s)) {
+				this.s = s;
+				this.outer.sendSubscription();
+				if (!hasInboundCancelled(this.state)) {
+					s.request(1);
+				}
+			}
+		}
+
+		@Override
+		public void onNext(T t) {
+			if (this.done) {
+				Operators.onNextDropped(t, currentContext());
+				return;
+			}
+
+			final T f = this.firstValue;
+			if (f == null) {
+				this.firstValue = t;
+
+				final long previousState = markFirstValueReceived(this);
+				if (hasInboundCancelled(previousState)) {
+					Operators.onDiscard(t, this.outer.currentContext());
+					return;
+				}
+
+				final Publisher<? extends R> result;
+				final CoreSubscriber<? super R> o = this.outer;
+
+				try {
+					result = Objects.requireNonNull(
+						this.transformer.apply(Signal.next(t, o.currentContext()), this),
+						"The transformer returned a null value"
+					);
+				}
+				catch (Throwable e) {
+					this.done = true;
+					o.onError(Operators.onOperatorError(this.s, e, t, o.currentContext()));
+					return;
+				}
+
+				result.subscribe(o);
+				return;
+			}
+
+			this.inner.onNext(t);
+		}
+
+		@Override
+		public void onError(Throwable t) {
+			if (this.done) {
+				Operators.onErrorDropped(t, this.outer.currentContext());
+				return;
+			}
+
+			this.done = true;
+			this.throwable = t;
+
+			final long previousState = markInboundTerminated(this);
+			if (hasInboundCancelled(previousState) || hasInboundTerminated(previousState)) {
+				Operators.onErrorDropped(t, this.outer.currentContext());
+				return;
+			}
+
+			if (hasFirstValueSent(previousState)) {
+				this.inner.onError(t);
+				return;
+			}
+
+			if (!hasFirstValueReceived(previousState)) {
+				final Publisher<? extends R> result;
+				final CoreSubscriber<? super R> o = this.outer;
+				try {
+					result = Objects.requireNonNull(this.transformer.apply(Signal.error(t,
+							o.currentContext()), this),
+							"The transformer returned a null value");
+				}
+				catch (Throwable e) {
+					o.onError(Exceptions.addSuppressed(e, t));
+					return;
+				}
+
+				result.subscribe(o);
+			}
+		}
+
+		@Override
+		public void onComplete() {
+			if (this.done) {
+				return;
+			}
+
+			this.done = true;
+
+			final long previousState = markInboundTerminated(this);
+			if (hasInboundCancelled(previousState) || hasInboundTerminated(previousState)) {
+				return;
+			}
+
+			if (hasFirstValueSent(previousState)) {
+				this.inner.onComplete();
+				return;
+			}
+
+			if (!hasFirstValueReceived(previousState)) {
+				final Publisher<? extends R> result;
+				final CoreSubscriber<? super R> o = this.outer;
+
+				try {
+					result =
+							Objects.requireNonNull(this.transformer.apply(Signal.complete(
```

Comment:
maybe make it a bit more readable by splitting creation of the `Signal` out on its own line/variable?